itest/rust/src/register_tests/func_test.rs

```diff
@@ -40,9 +40,71 @@ impl FuncRename {
     }
 }
 
+#[derive(GodotClass)]
+#[class(base=RefCounted)]
+struct GdSelfReference {
+    internal_value: i32,
+
+    #[base]
+    base: Base<RefCounted>,
+}
+
+#[godot_api]
+impl GdSelfReference {
+    // A signal that will be looped back to update_internal through gdscript.
+    #[signal]
+    fn update_internal_signal(new_internal: i32);
+
+    #[func]
+    fn update_internal(&mut self, new_value: i32) {
+        self.internal_value = new_value;
+    }
+
+    #[func]
+    fn fail_to_update_internal_value_due_to_conflicting_borrow(
+        &mut self,
+        new_internal: i32,
+    ) -> i32 {
+        // Because a self reference is held while the signal is emitted when
+        // gdscript tries to call update_internal there will be a failure due
+        // to the double borrow and self.internal_value won't be changed.
+        self.base.emit_signal(
+            "update_internal_signal".into(),
+            &[new_internal.to_variant()],
+        );
+        return self.internal_value;
+    }
+
+    #[func(gd_self)]
+    fn succeed_at_updating_internal_value(mut this: Gd<Self>, new_internal: i32) -> i32 {
+        // Because this isn't bound while the signal is emitted when gdscript
+        // will succeed at calling update_internal and self.internal_value will
+        // be changed.
```

Comment:
You don't use commas, which makes things very hard to read for me. I would suggest:
```suggestion
        // Since a self reference is held while the signal is emitted, when
        // GDScript tries to call update_internal(), there will be a failure due
        // to the double borrow and self.internal_value won't be changed.
        self.base.emit_signal(
            "update_internal_signal".into(),
            &[new_internal.to_variant()],
        );
        return self.internal_value;
    }

    #[func(gd_self)]
    fn succeed_at_updating_internal_value(mut this: Gd<Self>, new_internal: i32) -> i32 {
        // Since this isn't bound while the signal is emitted, GDScript will succeed at calling
        // update_internal() and self.internal_value will be changed.
```